src/utils/govPayMetadata.ts

```diff
@@ -0,0 +1,91 @@
+import { Passport } from "../models";
+import {
+  GovPayMetadata,
+  GovPayMetadataValue,
+  GovUKCreatePaymentPayload,
+  Passport as IPassport,
+  Value,
+} from "../types";
+
+type FormattedMetadata = NonNullable<GovUKCreatePaymentPayload["metadata"]>;
+
+const isPassportValue = (value: GovPayMetadataValue) =>
+  typeof value === "string" && value.startsWith("@");
+
+/**
+ * Convert GovPayMetadata set in Editor to format accepted by GovPay API
+ * Read dynamic data variables from passport and inject into output
+ */
+const parseMetadata = (
+  metadata: GovPayMetadata[],
+  passport: Passport,
+): FormattedMetadata => {
+  let entries: [string, GovPayMetadataValue][] = [];
+
+  entries = metadata.map(({ key, value }) => {
+    if (!isPassportValue(value)) return [key, value];
+
+    // Remove "@" prefix
+    const passportKey = (value as string).substring(1);
+    const passportValue = passport.any<Value>([passportKey]);
+
+    // Validate and format
+    const validatedValue = validateMetadata(key, passportValue);
+    return [key, validatedValue];
+  });
+
+  const parsedMetadata = Object.fromEntries(entries);
+  return parsedMetadata;
+};
+
+/**
+ * Validate that GovPayMetadata formatting rules are adhered to
+ * Initial validation is done in-Editor, this is intended to catch issues
+ * with dynamic passport variables
+ */
+const validateMetadata = (
+  key: string,
+  value: Value | undefined,
+): GovPayMetadataValue => {
+  let validated: GovPayMetadataValue | null = null;
+
+  switch (typeof value) {
+    case "string":
+    case "number":
+    case "boolean":
+      validated = value;
+      break;
+    case "object":
+      validated = Array.isArray(value)
+        ? value.join(", ")
+        : `Error: Invalid metadata value for "${key}" set in PlanX`;
+      break;
+    default:
+      validated = `Error: Invalid metadata value for "${key}" set in PlanX`;
+      break;
+  }
+
+  // Truncate to max 100 characters
+  // TODO: Should we truncate or throw error here?
+  if (typeof validated === "string") validated = validated.substring(0, 100);
```

Comment:
I'm actually more inclined to pass back an error here if the value exceeds the allowed length.

I suspect that an error is a probably more meaningful than a truncated value - maybe one for POs to answer?

Reply:
I'd actually vote truncate I think ! Just ideally in an obvious way (eg using `...`) 

If throwing an error, do you actually anticipate POs are ready to self-troubleshoot & fix or would this just immediately bounce up in help-issues? I'm (perhaps too negatively!) imagining the latter.

Reply:
Using an ellipsis is actually a very good idea - I'll make this change.

My reasoning for starting the string with `"Error:` would be that it should be immediately obvious to the person who has GovPay access that something is wrong here, whilst a truncated string might not obviously be wrong (maybe unlikely, but an array of values that ended up being cut to 100 characters without any obvious errors could not get spotted).

I think we go with an ellipsis for now, and wait for feedback (and also raise at PO testing). The use cases for this feature currently are to get a number only so this may not really be a relevant issue.

Reply:
Done! See [1923f12](https://github.com/theopensystemslab/planx-core/pull/348/commits/1923f12660404b90b6a56cca35127f71bb18cfaa)